Fix boost build on Windows

jamessmillie opened the pull request:
Boost, when compiling libraries, failed to build and install correctly on Windows. This is because the bootstrapping process on Windows takes different arguments and, therefore, the install command must be modified to include to build instructions that are passed in via `bootstrap.sh` on Linux/Mac systems. The format and exact arguments also differ on Windows. 

In this PR:

- the bootstrapping for Windows is moved to a new method
- the arguments for the Windows install command are updated to match the `b2.exe` help documentation

jamessmillie committed:
First attempt at fixing the boost build on Windows. The args are different than on unix and the bootstrapping process is also different, so need to adjust the bootstrap args and b2 install args accordingly.

jamessmillie committed:
Cant use --with-<library> and --without-<library> in the same command line, so just explicitly remove libraries and the rest will be built aautomatically

jamessmillie committed:
Some of the libs in all_libs are required and cannot be excluded. the address-model option was required

jamessmillie committed:
The runtime-link property can only be static or shared, not both

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Spack style fixes

jamessmillie committed:
Update wording in docstring

jamessmillie committed:
Cleaning up package file

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -728,9 +759,12 @@ def install(self, spec, prefix):
+        if self.spec.satisfies("platform=windows"):
+            b2_options = [
+                f"--prefix={self.prefix}",
+                f"address-model={64 if platform.machine().endswith('64') else 32}",
```
(question) Can this not be determined from `self.spec.target`?

jamessmillie replied:
Yes, it looks like `self.spec.target.family` contains this information.

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -696,16 +730,13 @@ def install(self, spec, prefix):
         if self.spec.satisfies("platform=windows"):
-            bootstrap = Executable("cmd.exe")
-            bootstrap("/c", ".\\bootstrap.bat", *bootstrap_options)
+            self.boostrap_windows()
```
Typo: boostrap -> bootstrap

(technically it's not a typo, because the function you are calling is in fact called `boostrap_windows`, but in both cases it should be changed to `bootstrap`)

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -652,6 +667,25 @@ def add_buildopt_symlinks(self, prefix):
+        if self.spec.satisfies("%msvc"):
+            # MSVC is the default compiler and is automatically detected by
+            # the bootstrapping script.
+            pass
```
Is it an error if we explicitly tell it to use `msvc`?

johnwparent replied:
See MSVC class API referencing MSVC toolset version: https://github.com/spack/spack/blob/develop/lib/spack/spack/compilers/msvc.py#L244

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -596,6 +597,20 @@ def determine_b2_options(self, spec, options):
+                # Without explicitly adding or removing specific libs, all would be built,
+                # so instead, explicitly exclude libs that aren't requested.
+                # Some items in all_libs cannot be excluded.
+                required_libs = ("signals",)
```
Can you add `conflicts(~signals, when="platform=windows")` as a class level directive (that should make this unnecessary)

jamessmillie replied:
Well, I don't think it's quite that simple. If every variant in `all_libs` is off, then only the headers are installed, and nothing is compiled. If you tell the package file that signals can't be shut off then you would never be allowed a "headers-only" install. But if you have other variants from within `all_libs` turned on (from my experience), the `signals` variant is required.

scheibelp replied:
> Well, I don't think it's quite that simple. If every variant in all_libs is off, then only the headers are installed, and nothing is compiled

That's a good point, but we don't want to have cases where we specified `~signals` but still have the `signals` library build.

I think the easiest way to do that is to force `+signals`; as you point out, that rules out some valid builds. You could also add a for loop among the library variants like:

```
for lib in [x, y, z]: # everything but signals
  conflicts(~signals, when=+lib)
```

Admittedly this applies a bit more care compared to other parts of the boost package, e.g. parts that remove items from `with_libs`: the section that declares variants would ideally include a version bound for those libraries that were introduced at some intermediate version of Boost.

Overall

* modifying the variant declarations doesn't need be done here
* but I'd still like more-careful handling of `+/~signals`

jamessmillie replied:
Ok, I have added conflicts for `~signals` when other lib variants are requested.

jamessmillie replied:
After further review, I needed to change `conflicts` to `requires` to get the expected behavior.

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -652,6 +667,25 @@ def add_buildopt_symlinks(self, prefix):
+            bootstrap_options.append("clang")
+
+        bootstrap = Executable("cmd.exe")
+        bootstrap("/c", ".\\bootstrap.bat", *bootstrap_options)
```
Things that `determine_bootstrap_options` was handling that I don't see handled here:

* `+python`
* `+icu`

Are those not supported on Windows? We may also want to add conflicts for those.

Overall it may be slightly better to make the flow consistent (right now the Windows path calls the `do-this` function, and the Linux path calls the `set-everything-up-for-me-and-then-I'll-do-it` function, I recommend they both follow the same high-level approach.

jamessmillie replied:
> Things that determine_bootstrap_options was handling that I don't see handled here:
> 
> +python
> +icu
>
> Are those not supported on Windows? We may also want to add conflicts for those.

* `icu` requires autotools, so I will add a conflict for windows
* `python` works, I will put in a couple of changes to get it working on windows


 
>Overall it may be slightly better to make the flow consistent (right now the Windows path calls the do-this function, and the Linux path calls the set-everything-up-for-me-and-then-I'll-do-it function, I recommend they both follow the same high-level approach.

Yes I agree, but I was trying to get Windows working without making any changes that might affect Linux users since Boost is such a fundamental package. The difference in bootstrapping between Linux and Windows is bizarre to me.

johnwparent replied:
@scheibelp do you want the `icu` support here? I have a patch that ports ICU to Windows that's been sitting in my QT port branch.

@jamessmillie this would require no extra work for you.

scheibelp replied:
> do you want the icu support here?

I'd prefer a later PR

@jamessmillie I see you added a `conflicts(+icu, when=windows)`, but does there also have to be a `--without-icu` argument passed to the bootstrap script?

jamessmillie replied:
No, I tried to add this flag and the build fails with it in the command line.

scheibelp reviewed (changes requested):
Additional requests

jamessmillie committed:
Using the spec target to determine if the system is 64 bit

jamessmillie committed:
Fix typo in the word 'bootstrap'

jamessmillie committed:
Explicitly setting msvc version for boost bootstrapping phase

jamessmillie committed:
Adding conflict for icu on windows due to autotools build system

jamessmillie committed:
Updating bjam file on windows and fixing +python variant configuration

jamessmillie committed:
Replace path separators on windows bjam file

jamessmillie committed:
Spack style fixes

alalazo committed:
Cherry-pick f242e0fd0cf2d5e26204aed0f780b85246416bb0

remove platform=cray (#43796)

Remove support for `cray` as a separate platform.

Any platform previously detected as `cray` is now detected as `linux`.

Users who still need platform=cray have to stick to Spack 0.22

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -597,6 +622,20 @@ def determine_b2_options(self, spec, options):
+                # Some items in all_libs cannot be excluded.
+                required_libs = ("signals",)
+                if f"+{lib}" not in spec and lib not in required_libs:
+                    options.append(f"--without-{lib}")
```
(question) does boost build every library unless it gets `--without-lib`? This differs from behavior on Linux (where we have to specify `--with-lib` to get any library.

jamessmillie replied:
Correct, when you build with no `--with-<opt>` or `--without-<opt>` flags, you build everything on windows. If you add a `--with-<opt>` flag, then you only get what you've requested, and if you use a `--without-<opt>` flag, then you build everything that isn't explicitly excluded.

Also important to note that you cannot mix `--with-` and `--without-` flags.

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -736,15 +789,24 @@ def install(self, spec, prefix):
+
+            b2_options = [f"--prefix={self.prefix}", f"address-model={64 if is_64bit() else 32}"]
+            if not self.spec.satisfies("+python"):
+                b2_options.append("--without-python")
```
Would it work to explicitly specify `--with-python` when `+python`, or does b2 not accept that on windows?

jamessmillie replied:
You cannot mix `--with-<opt>` and `--without-<opt>`, so since the rest of the package file uses `--without-<opt>`, then this must be `--without-python` if the spec does not satisfy `+python`

scheibelp replied:
(question) Is this a Windows-specific issue?

On Linux right now (with the `boost` Spack package from `develop`), if you do `spack install boost +python ~icu` it will pass `--with-python --without-icu` as options to the bootstrapping script.

jamessmillie replied:
Yes, appears to be a Windows difference which is that these options are initially provided during bootstrapping on linux, but on windows, the bootstrapping process does not accept the `--with` or `--without` flags, they are only provided to `b2`, which, when I tried to mix `with` and `without` flags, I got an error message on Windows.

scheibelp replied:
(question) So if I understand:

```
./b2 --without-python --without-icu #works
./b2 --with-python --without-icu # fails
./b2 --with-python # works, and doesn't use icu
```

Would you be able to include the error message here?

jamessmillie replied:
`error: both --with-<library> and --without-<library> specified`

scheibelp replied:
The wording of that error message makes it sound like it got something like `--with-python --without-python` (i.e. `--with-x` and `--without-x` where X is the same in both cases) - are you not seeing anything like that in the build log?

jamessmillie replied:
I agree the message wording is ambiguous, however, I get the same error message with a much simpler `b2` command where I mix one `--with` and one `--without` flag: 

```
.\b2.exe --with-atomic --without-chrono
error: both --with-<library> and --without-<library> specified
```

scheibelp replied:
Thanks for adding that!

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -525,10 +543,17 @@ def determine_bootstrap_options(self, spec, with_libs, options):
                 # that can be used as a compiler option instead.
                 mpi_line = "using mpi : %s" % spec["mpi"].mpicxx
+                if "platform=cray" in spec:
+                    mpi_line += " : <define>MPICH_SKIP_MPICXX"
```
(question) have you tested this on Cray? Most systems that are Crays should actually report Linux as the platform (e.g. `spack arch`).

jamessmillie replied:
This was the result of a bad conflict resolution. should be fixed now.

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -492,6 +495,18 @@ def bjam_python_line(self, spec):
+    def bjam_python_line_win(self, spec):
+        # avoid "ambiguous key" error
+        if spec.satisfies("@:1.58"):
+            return ""
```
Can you explain what this is? Do you think there is another mechanism that should be used to specify Python before `1.59`? Or perhaps we can just say that `require(@1.59:, when="platform=windows")`?

jamessmillie replied:
This is [in](https://github.com/spack/spack/blob/a7444873b9fa9a5837ba0a0853845b6691859051/var/spack/repos/builtin/packages/boost/package.py#L486) the regular `bjam_python_line` method that I copied , so I don't know why it was added originally. I can remove it if it is not useful.

scheibelp replied:
Ah ok I think if you move the `as_posix` calls to that method, you can reuse it on both systems instead of creating this secondary method (using `as_posix` should be fine on Linux). And would also resolve

```
spec["python"].prefix.libs.as_posix(),
```

(`.libs` is a container object, not a path, so you can't call `as_posix()` on it).

jamessmillie replied:
Ok I put them back into the same singular `bjam_python_line` method, but the actual paths are different on windows, so I just made the return conditional on OS instead.

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -492,6 +495,18 @@ def bjam_python_line(self, spec):
+            spec["python"].version.up_to(2),
+            spec["python"].command.path.replace("\\", "/"),
+            spec["python"].prefix.include.replace("\\", "/"),
+            spec["python"].prefix.libs.replace("\\", "/"),
```
(minor) I think it might be slightly more robust to do `pathlib.Path(...).as_posix`

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Using as_posix instead of replace

jamessmillie committed:
Making ~signals a conflict in the case where other libs are being built on windows

jamessmillie committed:
Explicitly remove icu from windows build since it is a conflict

jamessmillie committed:
This actually broke the build, so not necessary to exclude

jamessmillie committed:
Should use requires instead of conflicts to get expected behavior of enforcing +signals when requesting other libs

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -193,6 +193,9 @@ def libs(self):
     # 1.84.0 dropped support for 98/03
     conflicts("cxxstd=98", when="@1.84.0:")
 
+    # Autotools packages are not available on Windows
```
Thanks for the comment: we have a general constraint that prevents building autotools-based packages, so I would assume even if this constraint doesn't exist, `spack spec boost+icu` would fail: are you not seeing that?

I think this was done on account of https://github.com/spack/spack/pull/43732/files#r1579910150, but FWIW I was more referring to controlling that using options to bootstrapping/b2 (if possible, it sounds like it wasn't possible to use `--without-icu`).

jamessmillie replied:
Yes, concretization fails when you add `+icu` to boost (after I took out the conflict) due to autotools. Should I take out the conflict then since it is handled elsewhere?

scheibelp replied:
Yes - thanks!

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -527,10 +550,15 @@ def determine_bootstrap_options(self, spec, with_libs, options):
                 # that can be used as a compiler option instead.
                 mpi_line = "using mpi : %s" % spec["mpi"].mpicxx
+                if spec.satisfies("platform=windows"):
+                    mpi_line = mpi_line.replace("\\", "/")
```
(minor) using `pathlib.Path(spec["mpi"].mpicxx).as_posix()` would be clearer IMO.

scheibelp reviewed (changes requested):
Thanks for the updates! I have a few additional requests (generally minor, but I think one at least would cause problems building `boost` on Windows).

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Instead of having a separate bjam_python_line method for windows, just make the line in bjam_python_line conditional

jamessmillie committed:
Use as_posix instead of manually changing backslashes

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -490,12 +499,20 @@ def bjam_python_line(self, spec):
+                spec["python"].version.up_to(2),
+                spec["python"].command.path,
+                spec["python"].headers.directories[0],
+                spec["python"].libs[0],
```
The first two entries of each of these are the same (or in this second case I think you can use `spec["python"].command.path.as_posix()`).

The third item here would tend to be more robust as it's expressed in the non-windows case: does it return something that isn't usable for you on Windows (after converting to posix)?

The fourth item does seem significantly different.

Overall would it be possible to consolidate like:

```
if windows:
  include = 
else
  include = 
# Likewise for the fourth item
...
return "using python...".format(
  version,
  command.path.as_posix()
  include,
  ...
)
```

jamessmillie replied:
I was able to put that method back to essentially the original by fixing how the python package file looks for the libs directory on windows instead.

scheibelp reviewed (changes requested):
I have one request for consolidation of the `using python` line, and a question about `--with`/`--without` options provided to `b2`.

(Also fwiw leaving in the `conflicts(+icu)` seems reasonable to me for now and can be removed in a separate PR once `icu` is buildable on Windows)

johnwparent commented:
>  leaving in the conflicts(+icu) seems reasonable to me for now and can be removed in a separate PR once icu is buildable on Windows

https://github.com/spack/spack/pull/45547 is up and building on Windows, but I can see this is ready to go in. If we want I can rebase my branch on this one and  make the "remove `icu` conflict" change in my branch, just let me know how you both want to proceed.

scheibelp commented:
> https://github.com/spack/spack/pull/45547 is up and building on Windows,

Thanks!

> If we want I can rebase my branch on this one and make the "remove icu conflict" change in my branch, just let me know how you both want to proceed.

that is my preference

johnwparent commented:
> > If we want I can rebase my branch on this one and make the "remove icu conflict" change in my branch, just let me know how you both want to proceed.
> 
> that is my preference

Done!

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Eliminating duplicate definitions of 2 paths provided to b2/bjam

jamessmillie committed:
spack style

jamessmillie committed:
Rather than having 2 separate bjam lines in Boost, fix the finding of python libs dir on windows. Tested on Mac and Windows

scheibelp commented on var/spack/repos/builtin/packages/boost/package.py:
```diff
@@ -492,9 +501,9 @@ def bjam_python_line(self, spec):
-            spec["python"].libs[0],
+            Path(spec["python"].command.path).as_posix(),
+            Path(spec["python"].headers.directories[0]).as_posix(),
+            Path(spec["python"].libs[0]).parent.as_posix(),
```
(no change needed) I'll note this is a change to the behavior on linux: before it was taking the full path to the library including `libname.so`, but now provides the directory where the library exists.

The pipelines are succeeding on Linux with builds of boost, and this was originally introduced by https://github.com/spack/spack/commit/ce3ab503ded78d19aae45a938084d1f8670cc625, which focused more on Python than on boost, so I don't think it is a problem.

jamessmillie replied:
I tested on Mac prior to committing, too, and saw no difference. (even setting it to something crazy like `/tmp` made no difference for me), but was vital on Windows.

scheibelp commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -1066,10 +1066,14 @@ def libs(self):
+        # On Windows, we need to find the libs directory regardless of whether or
+        # not its a shared build, and .DLLs are next to the executable, so search
+        # for static first.
+        if not self.spec.satisfies("+shared") or self.spec.satisfies("platform=windows"):
```
The change you have here seems pretty close to what we want (good catch!) - the default `.libs` handler (in `spec.py`) specifically searches for `.lib` files regardless of whether the package was built `+shared` or not (because we decided that for windows, `.libs` is specifically for returning the link-time libraries). To that end, I think we actually want to drop `shared_libs` on windows and just search for static libs (I'll think more on this and state a preference tomorrow).

scheibelp replied:
To confirm, on Windows

* this should be just `static_libs`, not `static_libs + shared_libs`
* and it might be worth making a comment similar to what is in `_libs_default_handler` in `spec.py` (i.e. `.libs` is about libraries used for linking)
* and IMO it's worth raising an exception if none are found

jamessmillie replied:
An exception is already raised if none are found.

jamessmillie committed:
Removing +icu conflict since the concretizer will find the conflict automatically

scheibelp reviewed (changes requested):
Once https://github.com/spack/spack/pull/43732/files#r1725750463 is addressed, this will be ready (everything else looks ready).

scheibelp commented:
@spackbot run pipeline

jamessmillie committed:
On windows, we want the link libraries which are in the libs dir, so only search for static_libs

scheibelp commented:
@spackbot run pipeline

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Spack style

jamessmillie committed:
Putting determine bootstrap options back into if not windows block

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Merge branch 'develop' into bugfix/boost-on-windows

jamessmillie committed:
Make lib directory search platform-specific

hainest commented:
Sorry, I meant to jump in here before it got merged. I'm integrating this into #30627, but have a few questions.

```
# On Windows, the signals variant is required when building any of
# the all_libs variants.
for lib in all_libs:
    requires("+signals", when=f"+{lib} platform=windows")
```
There are two issues here:

1. `signals` is no longer available as of 1.68.0. It was replaced by `signals2` which has been available since 1.39.0. I _think_ we could just use `signals2` here, but I have no idea what the Windows support looks like.
2. 30627 replaces `all_libs` with conditional variants because it's incomplete, incorrect, and a maintenance hazard. That makes this logic questionable. Why are all of those libraries depending on the deprecated `signals` lib _only_ on Windows? Are there other libraries that require it? Maybe we can just conditionally turn on `signals{2}` when building on Windows. 
 
@scheibelp Is there a way to query spack for the OS at the point when a variant is being created, or should we use `sys.platform`?


---

```
for lib in self.all_libs:
    if f"+{lib}" not in spec:
        options.append(f"--without-{lib}")
```
Reading through your [comments](https://github.com/spack/spack/pull/43732/files#r1673059241), I _think_ I can handle this based on what libraries need to be compiled. I don't have a Windows machine to test on, so I might have to iterate on this with you.

scheibelp commented:
> signals is no longer available as of 1.68.0. It was replaced by signals2 which has been available since 1.39.0. I think we could just use signals2 here, but I have no idea what the Windows support looks like.

I think this could just be:

```
for lib in all_libs:
    requires("+signals", when=f"@:1.67 +{lib} platform=windows")
    # in the future, when a signals2 variant is added:
    requires("+signals2", when=f"@1.68: +{lib} platform=windows")
```

> 30627 replaces all_libs with conditional variants because it's incomplete, incorrect, and a maintenance hazard. That makes this logic questionable.

In your PR, you collected library name variants into `library_names`, I assume this loop could loop over that instead of `all_libs` to get the same effect (perhaps I'm overlooking a complication between the concepts of `when`/`buildable` in your PR though - I'm not clear what that distinction is used for).

(You should not have to add any more logic to account for the possibility of the variant being conditional: Spack will do that)

> Is there a way to query spack for the OS at the point when a variant is being created, or should we use sys.platform?

`sys.platform` is reasonable at variant creation time, spack also provides `spack.platforms.host().name`
